### PR TITLE
feat(box):  auth, sync users and delete user features

### PR DESCRIPTION
**Pull request by athielba**
## Description

Box complete setup
1. Authentication
2. Refresh  access token
3. Refresh access token before it expires
4. Sync user
5. Schedule sync users
6. Delete user

## Checklist:

Before you create this PR, confirm all the requirements listed below by checking the checkboxes like this (`[x]`).

- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests to cover the new feature or fixes.

### CODE REVIEW

File: apps/box/scripts/clean-database.sh [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
```
Comment (elbalexandre) [inline apps/box/scripts/clean-database.sh:1]: We probably don't need that for every integration.
I propose to keep integration simple now and we could add a shared script that do that later.

File: apps/box/src/app/auth/route.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+      region,
+      sourceId: env.ELBA_SOURCE_ID,
+      baseUrl: env.ELBA_REDIRECT_URL,
+      error: 'unauthorized',
+    });
+  }
+
+  await setupOrganisation({ organisationId, code, region });
+
+  redirect(env.ELBA_REDIRECT_URL, RedirectType.replace);
```
Comment (elbalexandre) [inline apps/box/src/app/auth/route.ts:27]:
```suggestion
  redirect(
      getRedirectUrl({
        region,
        sourceId: env.ELBA_SOURCE_ID,
        baseUrl: env.ELBA_REDIRECT_URL,
      })
  , RedirectType.replace);
```

File: apps/box/src/app/auth/service.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+      accessToken: encodedAccessToken,
+      refreshToken: encodedRefreshToken,
+      region,
+    })
+    .onConflictDoUpdate({
+      target: organisationsTable.id,
+      set: {
+        accessToken: encodedAccessToken,
+        refreshToken: encodedRefreshToken,
+        region,
```
Comment (elbalexandre) [inline apps/box/src/app/auth/service.ts:37]: My bad, I let that live in the template, but it's useless as both region have different DB.
```suggestion
```

File: apps/box/src/common/apiclient.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
+import { BoxError } from '@/connectors/commons/error';
```
Comment (elbalexandre) [inline apps/box/src/common/apiclient.ts:1]: As discussed, let's avoid this for now.

File: apps/box/src/database/schema.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import { uuid, text, pgTable } from 'drizzle-orm/pg-core';
+import { type InferSelectModel } from 'drizzle-orm';
+
+export const organisationsTable = pgTable('organisations', {
+  id: uuid('id').primaryKey(),
+  region: text('region').notNull(),
+  accessToken: text('accessToken').notNull(),
+  refreshToken: text('refreshToken').notNull(),
+});
```
Comment (elbalexandre) [inline apps/box/src/database/schema.ts:9]:
```suggestion
  createdAt: timestamp('created_at').default(sql`now()`).notNull(),
});
```

File: apps/box/src/inngest/functions/users/delete-users.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import { decrypt } from '@/common/crypto';
+
+export const deleteSourceUsers = inngest.createFunction(
+  {
+    id: 'box-delete-users',
+    concurrency: {
+      key: 'event.data.organisationId',
+      limit: 1,
+    },
+    retries: 5,
```
Comment (elbalexandre) [inline apps/box/src/inngest/functions/users/delete-users.ts:16]: Let's put the 2 values in env var and bump concurrency number (we need to check the doc for RateLimit but we could go higher than that - 5 seems acceptable)

File: apps/box/src/inngest/functions/users/schedule-users-synchronize.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+export const scheduleUsersSynchronize = inngest.createFunction(
+  { id: 'box-schedule-users-syncs' },
+  { cron: env.USERS_SYNC_CRON },
+  async ({ step }) => {
+    const organisations = await db
+      .select({
+        id: organisationsTable.id,
+        region: organisationsTable.region,
+        accessToken: organisationsTable.accessToken,
+        refreshToken: organisationsTable.refreshToken,
```
Comment (elbalexandre) [inline apps/box/src/inngest/functions/users/schedule-users-synchronize.ts:15]:
```suggestion
```

File: apps/box/src/inngest/functions/users/synchronize-users.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+  {
+    id: 'box-synchronize-users',
+    priority: {
+      run: 'event.data.isFirstSync ? 600 : -600',
+    },
+    concurrency: {
+      key: 'event.data.organisationId',
+      limit: 1,
+    },
+    retries: 3,
```
Comment (elbalexandre) [inline apps/box/src/inngest/functions/users/synchronize-users.ts:30]: Same here

File: apps/box/src/inngest/middlewares/rate-limit-middleware.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+                retryAfter = parseInt(retryAfterHeader, 10);
+              }
+
+              return {
+                ...context,
+                result: {
+                  ...result,
+                  error: new RetryAfterError(
+                    `API rate limit reached by '${fn.name}', retry after ${retryAfter} seconds.`,
+                    retryAfter.toString(),
```
Comment (elbalexandre) [inline apps/box/src/inngest/middlewares/rate-limit-middleware.ts:48]: We should put the time indicator, if we return a number it's parsed at millisecond, here it's seconds.
```suggestion
                    `${retryAfter.toString()}s`,
```

Comment (athielba) [reply apps/box/src/inngest/middlewares/rate-limit-middleware.ts:48]: Good catch

Comment (athielba) [reply apps/box/src/inngest/middlewares/rate-limit-middleware.ts:48]: I will use `ms` as others

File: apps/box/src/common/apiclient.ts (hunk 4 above)
Comment (athielba) [reply apps/box/src/common/apiclient.ts:1]: Maksym is using the same  type of Error class everywhere , I don't think we discussed about this  (We  discussed about  general rest client pattern., isn't it ?)

Comment (elbalexandre) [reply apps/box/src/common/apiclient.ts:1]: I was talking about the apiclient not the error

File: apps/box/src/database/schema.ts (hunk 5 above)
Comment (elbalexandre) [reply apps/box/src/database/schema.ts:9]: It's convenient to debug issues in prod.

Comment (athielba) [reply apps/box/src/database/schema.ts:9]: Are we going to apply this changes everywhere ?

**Review by proquig — APPROVED** [review]

**Pull request merged**